text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,289 @@
+
+#### Building of `compiler-builtins`
+
+Currently, `compiler-builtins` contains components implemented in the C programming language. While these dependencies have been highly optimized, the use of them would require the builder of the root crate to also have a sane compilation environment for compilation in C.
+
+This RFC proposes instead to use the [pure rust implementation] when compiling for a custom target, removing the need for a C compiler.
+
+While this may have code size or performance implications, this would allow for maximum portability.
+
+[pure rust implementation]: https://github.com/rust-lang-nursery/compiler-builtins
+
+#### `RUSTC_BOOTSTRAP`
+
+It is necessary to use unstable features to build `core`. In order to allow users of a stable compiler to build `core`, we would set the `RUSTC_BOOTSTRAP` environment variable **ONLY** for the compilation of `core`.
+
+This should be considered sound, as stable users may not change the source used to build `core`, or the features used to build `core`.
+
+## 2 - Introduce the concept of "stable features" for `core`
+
+### Use Case
+
+In some cases, it may be desirable to modify `core` in set of predefined manners. For example, on some targets it may be necessary to have lighter weight machinery for `fmt`.
+
+This step would provide a path for stabilization of compile time `core` features, which would be a subset of all total compile time features of `core`.
+
+### Caveats
+
+Initially, the list of stable compile time features for `core` would be empty, as none of the current features have had an explicit decision to be stable or not.
+
+### User Interaction
+
+Compile time features for `core` may be specified using the same Cargo.toml syntax used for other crates.
+
+The syntax is the same when using `unstable` and `stable` features, however the former may only be used with a nightly compiler, and use of an `unstable` feature with a stable compiler would result in a compile time error.
+
+The syntax for these features would look as follows:
+
+```toml
+[dependencies.core]
+default-features = false
+features = [...]
+```
+
+It is not necessary to explicitly mentioned the dependency of `core`, unless changes to features are necessary.
+
+### Technical Implications
+
+#### Path to stabilization
+
+The stabilization of a `core` feature flag would require a process similar to the stabilization of a feature in the language:
+
+* Any new feature begins as unstable, requiring a nightly compiler
+* When the feature is sufficiently explored, an RFC/PR can be made to `libcore` to promote this feature to stable
+* When this has been accepted, the feature of `core` may be used with the stable compiler.
+
+#### Implementation of Stable Features
+
+There would be some mechanism of differentiating between flags used to build core, sorting them into the groups `unstable` and `stable`. This RFC does not prescribe a certain way of implementation.
```

Comment:
That's fair; but you may want to suggest a few possible mechanisms to make this implementable.

Reply:
Perhaps! But I am not an active developer of `core` or `std`, so I was trying to avoid something that makes sense to me, but doesn't make sense in practice.

Reply:
Fair :) Maybe someone else can help out here? e.g. @alexcrichton?